### PR TITLE
[Splunk] Use json responses instead XML

### PR DESCRIPTION
**Pull request by shmuel44**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: [link to the issue
](https://jira-hq.paloaltonetworks.local/browse/CIAC-6263)
## Description
- adding argument to splunk-sdk client, output_mode='json' instead to get default XML response .
- replacing ResultsReader (deprecated) to JSONResultsReader.
-  added some code improvements.
- added an error message when the `kv-store-collection-create` command tries to create an existing collection.

### CODE REVIEW

File: Packs/SplunkPy/ReleaseNotes/3_0_21.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+
+#### Integrations
+
+##### SplunkPy
+
+
```
Comment (dorschw) [inline Packs/SplunkPy/ReleaseNotes/3_0_21.md:6]: Add a RN

File: Packs/SplunkPy/pack_metadata.json [hunk 2]
```diff
@@ -2,7 +2,7 @@
     "name": "Splunk",
     "description": "Run queries on Splunk servers.",
     "support": "xsoar",
-    "currentVersion": "3.0.20",
+    "currentVersion": "3.0.21",
```
Comment (dorschw) [inline Packs/SplunkPy/pack_metadata.json:5]: this one is worth a `3.1.0`

File: Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py [hunk 3]
```diff
@@ -348,12 +348,13 @@ def fetch_notables(service: client.Service, mapper: UserMappingObject, cache_obj
     # if last_run_latest_time is not None it's mean we are in a batch fetch iteration with offset
     latest_time = last_run_latest_time or now
     kwargs_oneshot = build_fetch_kwargs(dem_params, occured_start_time, latest_time, search_offset)
+    kwargs_oneshot["output_mode"] = "json"
```
Comment (ilappe) [inline Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py:351]: pleas put it in `build_fetch_kwargs`

File: Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py [hunk 4]
```diff
@@ -1308,7 +1321,11 @@ def get_modified_remote_data_command(service: client.Service, args):
              '| fields rule_id ' \
              '| dedup rule_id'.format(last_update_splunk_timestamp)
     demisto.debug(f'Performing get-modified-remote-data command with query: {search}')
-    for item in results.ResultsReader(service.jobs.oneshot(search, count=MIRROR_LIMIT)):
+
+    for item in results.JSONResultsReader(service.jobs.oneshot(search, count=MIRROR_LIMIT, output_mode='json')):
+        if isinstance(item, results.Message):
+            demisto.debug(f"Message from splunk-sdk message: {item.message}")
+            continue
```
Comment (ilappe) [inline Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py:1328]: you can define a generic function like:
```
    def handle_messgae(item):
            if isinstance(item, results.Message):
                demisto.debug(f"Message from splunk-sdk message: {item.message}")
                return True
            return False
```
and then use it in all the relevant places 
```
        if handle_message(item):
            continue
```

File: Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py [hunk 5]
```diff
@@ -2394,9 +2421,18 @@ def replace_keys(data):
...
+        service.kvstore.create(args['kv_store_name'])
+    except HTTPError as error:
+        if error.status == 409 and error.reason == 'Conflict':
+            raise DemistoException(
+                f"KV store collection {service.namespace['app']} already exists.",
+            ) from error
+
+    return CommandResults(
+        readable_output=f"KV store collection {service.namespace['app']} created successfully",
+    )
```
Comment (ilappe) [inline Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py:2435]: looks good! 
but from where is it?

**Review by ilappe — CHANGES_REQUESTED** [review]
Nice job!

File: Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py (hunk 5 above)
Comment (shmuel44) [reply Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py:2435]: when i trying this command (create collection that already exists ),
 i got unclear error, so , i added it...

**Review by ilappe — APPROVED** [review]
great!

File: Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py [hunk 6]
```diff
@@ -5,18 +5,17 @@
...
+
 from splunklib import client
 from splunklib import results
 from splunklib.data import Record
-import urllib3
 from splunklib.binding import AuthenticationError, HTTPError, namespace
 
-urllib3.disable_warnings()
 
+OUTPUT_MODE = 'json'  # type of response from splunk-sdk query (json/csv/xml)
```
Comment (dorschw) [inline Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py:18]: rename to `OUTPUT_MODE_JSON`, to give future maintainers some idea about what it is without having to check its value

Comment (shmuel44) [reply Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py]: Done.

File: Packs/SplunkPy/Integrations/SplunkPy/SplunkPy_test.py [hunk 7]
```diff
@@ -1007,12 +1009,60 @@ def __init__(self):
...
+        "close_extra_labels": ["Custom"],
+        "mapper": splunk.UserMappingObject(Service(), False),
+    }
+    debug_mock = mocker.patch.object(demisto, "debug")
+    mocker.patch.object(demisto, "params", return_value={"timezone": "0"})
+    mocker.patch.object(demisto, "info")
+    mocker.patch(
+        "SplunkPy.results.JSONResultsReader", return_value=[Message("test message")]
+    )
+    mocker.patch.object(demisto, "results")
```
Comment (dorschw) [inline Packs/SplunkPy/Integrations/SplunkPy/SplunkPy_test.py:1052]: If we need this mock, let's check the arguments it's called with. 
If we don't, let's delete it

File: Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py [hunk 8]
```diff
@@ -2193,30 +2213,30 @@ def splunk_get_indexes_command(service: client.Service):
...
         data_formatted = data.encode('utf8')
-        r = index.submit(data_formatted, sourcetype=demisto.args()['sourcetype'], host=demisto.args()['host'])
-        demisto.results('Event was created in Splunk index: ' + r.name)
+        r = index.submit(data_formatted, sourcetype=args['sourcetype'], host=args['host'])
+        demisto.results(f'Event was created in Splunk index: {r.name}')
 
 
 def splunk_submit_event_hec(
-    hec_token: str,
+    hec_token: str | None,
```
[the remark this answers is not in the document]
Comment (shmuel44) [reply Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py]: manual 😄

File: Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py [hunk 9]
```diff
@@ -303,14 +301,15 @@ def build_fetch_kwargs(params, occured_start_time, latest_time, search_offset):
...
     }
 
 
 def build_fetch_query(params):
     fetch_query = params['fetchQuery']
 
-    if demisto.get(params, 'extractFields'):
-        extractFields = params['extractFields']
+    if (extract_fields := params.get('extractFields')):
+        extractFields = extract_fields
```
Comment (dorschw) [inline Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py:312]: a second variable? 🧐  

```suggestion
        extra_raw_arr = extract_fields.split(',')
```

File: Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py [hunk 10]
```diff
@@ -1644,18 +1659,19 @@ def readinto(self, b):
...
 
-    reader = results.ResultsReader(oneshotsearch_results)
+    reader = results.JSONResultsReader(oneshotsearch_results)
     for item in reader:
-        if isinstance(item, results.Message):
-            return item.message["clock"]
+        if handle_message(item):
+            continue
         if isinstance(item, dict):
             return item["clock"]
```
Comment (dorschw) [inline Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py:1673]: earlier we first check if the type is dict, then message. Let's keep the same logic here, unless there's a reason

File: Packs/SplunkPy/Integrations/SplunkPy/SplunkPy_test.py
Comment (dorschw) [inline Packs/SplunkPy/Integrations/SplunkPy/SplunkPy_test.py:1]: great job, please add a unit test specifically for `handle_message`

**Review by dorschw — APPROVED** [review]

File: Packs/SplunkPy/Integrations/SplunkPy/SplunkPy_test.py
Comment (shmuel44) [reply Packs/SplunkPy/Integrations/SplunkPy/SplunkPy_test.py:1]: done

File: Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py [hunk 11]
```diff
@@ -303,15 +301,15 @@ def build_fetch_kwargs(params, occured_start_time, latest_time, search_offset):
...
 
 def build_fetch_query(params):
     fetch_query = params['fetchQuery']
 
-    if demisto.get(params, 'extractFields'):
-        extractFields = params['extractFields']
-        extra_raw_arr = extractFields.split(',')
+    if (extract_fields := params.get('extractFields')):
+        extra_raw_arr = extract_fields.split(',')
         for field in extra_raw_arr:
```
Comment (dorschw) [inline Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py:313]:
```suggestion
        for field in extract_fields.split(','):
```

**Review by dorschw — APPROVED** [review]

File: Packs/SplunkPy/ReleaseNotes/3_1_0.md [hunk 12]
```diff
@@ -0,0 +1,8 @@
+
+#### Integrations
+
+##### SplunkPy
+
+- Improved implementation of integration commands response processing, XML handling has been deprecated and replaced by json.
```
Comment (ShirleyDenkberg) [inline Packs/SplunkPy/ReleaseNotes/3_1_0.md:6]:
```suggestion
- Improved implementation of integration commands response processing. XML handling has been deprecated and replaced by JSON.
```

**Comment by ShirleyDenkberg** [comment]
@dorschw @ilappe Doc review completed.

**Pull request merged**